### PR TITLE
fast-math: Fold float multiplication by minus one to unary negative

### PR DESCRIPTION
**Pull request by MaxGraey**
After #3155 we could add more NaN-sign sensitive simplifications for float operations.
`x * -1.0` -> `-x` (fneg(x))

### CODE REVIEW

File: src/passes/OptimizeInstructions.cpp [hunk 1]
```diff
@@ -1417,8 +1417,7 @@ struct OptimizeInstructions
     }
     {
       double value;
-      if (fastMath &&
-          matches(curr, binary(Abstract::Sub, any(), fval(&value))) &&
+      if (matches(curr, binary(Abstract::Sub, any(), fval(&value))) &&
```
Comment (MaxGraey) [inline src/passes/OptimizeInstructions.cpp:1420]: Also remove `fastMath` guard from here to allow canonicalize `x - (-0.0)` ==> `x + 0.0` which is always safe

**Comment by MaxGraey** [comment]
Fuzzed:
```
Invocations so far:
   FuzzExec: 14258
   CompareVMs: 3885
   CheckDeterminism: 1198
   Wasm2JS: 3382
   Asyncify: 3651

ITERATION: 16749
```

File: src/passes/OptimizeInstructions.cpp (hunk 1 above)
Comment (tlively) [reply src/passes/OptimizeInstructions.cpp:1420]: @kripken what was the verdict here? I thought we had decided this wasn't safe.

File: src/passes/OptimizeInstructions.cpp [hunk 2]
```diff
@@ -1440,10 +1439,14 @@ struct OptimizeInstructions
...
     }
+    // x * -1.0   ==>   -x
+    if (fastMath && matches(curr, binary(Abstract::Mul, any(), fval(-1.0)))) {
+      return builder.makeUnary(Abstract::getUnary(type, Abstract::Neg), left);
+    }
     if (matches(curr, binary(Abstract::Mul, any(&left), constant(1))) ||
         matches(curr, binary(Abstract::DivS, any(&left), constant(1))) ||
         matches(curr, binary(Abstract::DivU, any(&left), constant(1)))) {
-      if (curr->type.isInteger() || fastMath) {
+      if (fastMath || curr->type.isInteger()) {
```
Comment (tlively) [inline src/passes/OptimizeInstructions.cpp:1449]: I don't think changing this line helps with readability.

**Review by tlively — COMMENTED** [review]
The transformation itself looks good, but I'm not so sure about the other changes.

**Review by tlively — APPROVED** [review]
@MaxGraey, ping me once you've double checked the safety of `x - (-0.0) ==> x + 0.0` on V8 as @kripken suggested and we can merge this 👍

**Comment by MaxGraey** [comment]
This is WebAssembly fiddle which compare different variant of subtractions NaN (quiet, signal, arithm, non-canonical, signed and unsigned) and negative zero with similar but optimized version: https://webassembly.studio/?f=fs4kfeplvgb

On Chrome 85. All tests valid.
On Firefox 81. All tests valid as well.

**Pull request merged**